### PR TITLE
[PyCDE] If-Then-Else syntactic sugar for muxes

### PR DESCRIPTION
**Pull request by teqdruid**
Allows specifying muxes with `if`-like syntax. This is the closest we can
get without having to do AST inspection (I think).

```
  @module
  class IfDemo:
    cond = Input(types.i1)
    out = Output(types.i8)

    @generator
    def build(ports):
      with If(ports.cond):
        v = types.i8(1)
      with Else:
        v = types.i8(0)
      EndIf()
      ports.out = v
```

### CODE REVIEW

**Comment by teqdruid** [comment]
Missing code docs and tests. @mikeurbach @mortbopet WDYT?

**Comment by teqdruid** [comment]
Inspired by https://github.com/l74d/scoping

**Review by mortbopet — COMMENTED** [review]
Looks like a good first stab. If it's going to be checked into tree, i'd like to see a few more comments to explain some of the magic around using the stack (e.g. the techniques described [here](https://github.com/l74d/scoping/blob/main/scoping.py)). 

I might be missing it reading the implementation, but does this support
- References to nested assignments (first example below)
- Default assignments (second example)? (and by extension, partial assignments)

If we're able to capture those two cases then we can probably capture _most_ of what is done inside `always_comb` blocks.

Nested assignments:
```python
  @generator
  def build(ports):
    w = ports.a * ports.b
    with If(ports.cond):
      with Then:
        with If(ports.cond2):
          with Then:
            v = ports.a * ports.a
            u = v * ports.b
          with Else:
            v = ports.a * ports.b
            u = v * ports.b
      with Else:
        v = ports.b.as_uint(16)
        u = v * ports.a
```
Default assignments:
```python
  @generator
  def build(ports):
    w = ports.a * ports.b
    v = ports.a
    u = ports.b
    with If(ports.cond):
      with Then:
        with If(ports.cond2):
          with Then:
            v = ports.a * ports.a
            u = v * ports.b
      with Else:
        v = ports.b.as_uint(16)
        u = v * ports.a
```

**Comment by teqdruid** [comment]
@mortbopet Yes to both of the cases you mentioned. I've added them to the tests.

**Review by mortbopet — APPROVED** [review]
Code LGTM, feel free to merge with or without the nits.

File: frontends/PyCDE/test/behavioral.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# RUN: %PYTHON% py-split-input-file.py %s | FileCheck %s
```
[the remark this answers is not in the document]
Comment (teqdruid) [reply frontends/PyCDE/test/behavioral.py:1]: Yeah, I agree. That's my bad. I keep forgetting the standard. Thanks for keeping me honest.

File: frontends/PyCDE/src/pycde/behavioral.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+# it.
+_current_if_stmt = ContextVar("current_pycde_if_stmt")
+
+# Note: these constructs play around with Python stack frames. They use a Python
+# C function to flush the changes back to the interpreter.
+# PyFrame_LocalsToFast(...) is specific to CPython and as such, is
+# implementation-defined. So it _could_ change in subsequent versions of Python.
+# It is, however, commonly used to modify stack frames so would break many users
+# if it were changed. (So it probably won't be.) Tested with all the Python
+# versions for which we produce wheels on PyPI.
```
[the remark this answers is not in the document]
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/behavioral.py:25]: I don't think that's necessary. Our release pipeline (and any reasonable one) runs the tests which exercise this functionality. What would we do manually which the tests wouldn't catch? (And why couldn't we automate it and add it to the tests?)

**Comment by mikeurbach** [comment]
I think this is neat, but can you add more description to the docstring?

> Syntactic sugar for creation of muxes with if-then-else-ish behavioral
  syntax.

In these more complex, nested cases, how are the muxes created? I think I understand it from reading the code, but I'm thinking about whoever is going to use this. It might seem like this will just lower to SV if/else constructs, but that's not the same semantics as mux (at least w.r.t. X, potentially other things).

**Comment by teqdruid** [comment]
We'll be doing this as a part of https://github.com/llvm/circt/issues/3655 . Considering this is still subject to change, I don't really want to document it yet... but I've been saying that about most of PyCDE thus far. (Which either implies that I'm lazy or not satisfied with PyCDE... I suspect it's a combination of both.

**Comment by mikeurbach** [comment]
👌 i had missed that ticket, sounds good to me.

**Pull request merged**